tests/bluetooth/tester/src/audio/btp_bap_audio_stream.c

```diff
@@ -2,188 +2,234 @@
-	struct btp_bap_audio_stream *stream;
-	struct bt_bap_stream *bap_stream;
-	struct k_work_delayable *dwork;
-	struct net_buf *buf;
-	uint32_t size;
-	uint8_t *data;
-	int err;
-
-	dwork = k_work_delayable_from_work(work);
-	stream = CONTAINER_OF(dwork, struct btp_bap_audio_stream, audio_send_work);
-	bap_stream = audio_stream_to_bap_stream(stream);
+	NET_BUF_POOL_FIXED_DEFINE(tx_pool, CONFIG_BT_ISO_TX_BUF_COUNT,
+				  BT_ISO_SDU_BUF_SIZE(CONFIG_BT_ISO_TX_MTU),
+				  CONFIG_BT_CONN_TX_USER_DATA_SIZE, NULL);
+
+	/* This loop will attempt to send on all streams in the streaming state in a round robin
+	 * fashion.
+	 * The TX is controlled by the number of buffers configured, and increasing
+	 * CONFIG_BT_ISO_TX_BUF_COUNT will allow for more streams in parallel, or to submit more
+	 * buffers per stream.
+	 * Once a buffer has been freed by the stack, it triggers the next TX.
+	 */
+	while (true) {
+		int err = -ENOEXEC;
+
+		for (size_t i = 0U; i < ARRAY_SIZE(tx_streams); i++) {
+			struct bt_bap_stream *bap_stream = tx_streams[i].bap_stream;
+			const bool is_streaming = stream_is_streaming(bap_stream);
+			struct net_buf *buf;
+
+			if (!is_streaming ||
+			    atomic_get(&tx_streams[i].enqueued) >= MAX_ENQUEUE_CNT) {
+				continue;
+			}
 
-	if (stream->last_req_seq_num % 201 == 200) {
-		err = bt_bap_stream_get_tx_sync(bap_stream, &info);
-		if (err != 0) {
-			LOG_DBG("Failed to get last seq num: err %d", err);
-		} else if (stream->last_req_seq_num > info.seq_num) {
-			LOG_DBG("Previous TX request rejected by the controller: requested seq %u,"
-				" last accepted seq %u", stream->last_req_seq_num, info.seq_num);
-			stream->last_sent_seq_num = info.seq_num;
-		} else {
-			LOG_DBG("Host and Controller sequence number is in sync.");
-			stream->last_sent_seq_num = info.seq_num;
-		}
-		/* TODO: Synchronize the Host clock with the Controller clock */
-	}
+			buf = net_buf_alloc(&tx_pool, K_SECONDS(1));
+			__ASSERT(buf != NULL, "Failed to get a TX buffer");
 
-	/* Get buffer within a ring buffer memory */
-	size = ring_buf_get_claim(&audio_ring_buf, &data, bap_stream->qos->sdu);
-	if (size > 0) {
-		buf = net_buf_alloc(&tx_pool, K_NO_WAIT);
-		if (!buf) {
-			LOG_ERR("Cannot allocate net_buf. Dropping data.");
-		} else {
 			net_buf_reserve(buf, BT_ISO_CHAN_SEND_RESERVE);
-			net_buf_add_mem(buf, data, size);
-
-			/* Because the seq_num field of the audio_stream struct is atomic_val_t
-			 * (4 bytes), let's allow an overflow and just cast it to uint16_t.
-			 */
-			stream->last_req_seq_num = (uint16_t)atomic_get(&stream->seq_num);
 
-			LOG_DBG("Sending data to stream %p len %d seq %d", bap_stream, size,
-				stream->last_req_seq_num);
+			net_buf_add_mem(buf, btp_bap_audio_stream_mock_data, bap_stream->qos->sdu);
+
+			err = bt_bap_stream_send(bap_stream, buf, tx_streams[i].seq_num);
+			if (err == 0) {
+				tx_streams[i].seq_num++;
+				atomic_inc(&tx_streams[i].enqueued);
+			} else {
+				if (!is_streaming) {
```

Comment:
this is const and already checked above,   should this be open coded to stream_is_streaming() instead of const bool?

Reply:
Ah, yes, bad optimization from my side :D Nice catch